ENH: added empty input handling `scipy.moment`; doc enhancements

aeklant opened the pull request:
closes gh-591

aeklant commented:
Please review, especially the documentation and formula in LaTeX format. 

There is a reference non-compliant with PEP8 but since this is a link I thought it should stay that way.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1017,8 +1017,11 @@ def moment(a, moment=1, axis=0):
+    If the points represent probability density, then the zeroth moment is the
+    total probability (i. e. one), the first moment is the mean, the second
+    moment is the variance, the third moment is the skewness and the fourth
+    moment divided by the square of the variance is the kurtosis.
```
I'm not sure about this paragraph. First central moment is not the mean, it's zero. Skewness is typically defined to be scale-free, so it's not just 3rd central moment. And I'm not sure there's a need to bring any of this into the docstring. 

I'd suggest to just stress that this a central moment (maybe, an example) and add an explicit formula.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1037,8 +1040,32 @@ def moment(a, moment=1, axis=0):
+    -----
+    [1]_ The n-th moment of a real-valued continuous function f(x) of a real
+    variable about a value c is:
+        \mu _{n} = \int_{-\infty }^{\infty }(x - c)^n  f(x) dx
```
Since you asked to review the the TeX notation: 

* it needs to have blanks line above and below.
* it needs to have the reST magic to render: `.. math::`  (notice the space between after the dots)

I personally prefer to minimize the number of curly brackets (here: `\infty` is a single symbol for TeX, hence `_{\infty}` and `_\infty` are the same for TeX). Consistent spacing helps readability, too. Here's how I'd write it:

```

.. math::

    \mu_n  = \int_{-\infty}^\infty (x - c)^n f(x) dx
```

(When typesetting, you'd add `\,` before dx, but it clutters the plain text version, so don't).

In any case, this formula should not be here --- better add a formula for what is actually being calculated.

rgommers replied:
I agree. So the formula should be a finite sum instead of an infinite integral. Here's a formula: http://www.r-tutor.com/elementary-statistics/numerical-measures/moment

aeklant commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+
+def _moment(a, moment, axis):
+    if not issubclass(type(moment), int):
+        raise ValueError("All moment parameters must be integers")
```
I added this because it was addressed in the comments, although type checking doesn't seem very pythonic; there might be a better way to handle this. Input appreciated.

rgommers replied:
This won't work well:

    In [7]: issubclass(np.int32, int)
    Out[7]: False

If you want to do a type check, this will work:

    np.issubdtype(dt, np.integer)  # works for all Python and Numpy integer types

Another and maybe simpler way to do it that would also work for floats that are integer-valued (so 1.0 but not 1.2):

    if not np.abs(moment - np.round(moment)) > 0:
        # Accept ints and integer-valued floats

aeklant commented:
A few notes on the latest commit (`bbdc0f9`):
* The reason for the addition of `_moment` is due to the fact that the recursive calls generated a problem with name collision (between the parameter `moment` and the function `moment`). I could change the name of the `moment` parameter but this could break some code so I opted for the helper function option.
* `moment == 0` and `moment == 1` have some repeated lines so I could have fused them into one `if` statement but the function has enough nested statements so I didn't want to add more.
* I made a few changes to the documentation. I hope it is clearer now; if not, let me know.
* I added a test for `moment == 0` and a test when requesting several moments.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+    Notes
+    -----
+    [1]_ The n-th moment of a real-valued continuous function f(x) of a real
+    variable about a value c is:
```
Still need a blank line here for it to render correctly.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+    variable about a value c is:
+    .. math::
+
+    m_k = \frac{1}{n} \sum_{i = 1}^n (x_i - \bar{x})^k
```
And this needs to be indented (4 spaces)

ev-br replied:
Some proofreading would be helpful: n-th moment or m_k, sum over I or k or m, "continuous function f(x)" has nothing to do with anything here, what is x-bar and what is a "value c".

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Moment_(mathematics)
```
Uhm, FWIW I don't think this wikipedia reference adds much value?

rgommers commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+        del shape[axis]
+        if shape:
+            # return an actual array of the appropriate shape
+            return np.zeros(shape, dtype=float) * 1.0
```
This isn't right (will be 0's), should be ``np.ones(shape, dtype=float)``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,58 @@ def moment(a, moment=1, axis=0):
+            return np.zeros(shape, dtype=float) * 1.0
+        else:
+            # the input was 1D, so return a scalar instead of a rank-0 array
+            return np.float64(1.0)
```
This can simply be ``1.0`` instead of ``np.float64(1.0)``

rgommers commented:
@aeklant the handling of ``array_like`` values for ``moment`` looks OK to me, but I would combine the ``moment == 0`` and ``moment == 1`` branches to cut down on code duplication.

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1480,6 +1482,16 @@ def test_moment(self):
+        assert_approx_equal(y[0], 0.0, 10)
+        assert_approx_equal(y[1], 1.25)
+        assert_approx_equal(y[2], 0.0)
+        assert_approx_equal(y[3], 2.5625)
```
Maybe add a comment above this "check array_like moment", and the checks can do done at once: ``assert_allclose(y, [0, 1.25, 0, 2.5625], rtol=xxx)``

aeklant commented:
@rgommers the reasons why I didn't combine them was because it only saves a couple of lines (I would still have to check around `if shape` to know whether to return zeros or ones) and I would then have to add a nested check. 

In short I opted to sacrifice saving those two or three lines of code in order to avoid making the function more complex and difficult to read. I am of course open to changing it if you still think it would be better.

aeklant commented:
@rgommers is there a downside to:

`if moment - np.round(moment) != 0`
instead of:
`if np.abs(moment - np.round(moment)) > 0`

I tested it and it seems to work the same and the call to `np.abs` would not be necessary. I just want to make sure I'm not missing something since floats are tricky sometimes.

aeklant commented:
I changed the way the integer check is done as discussed and added tests for it. I also tried to do a better job with the documentation.

Please review.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1015,16 +1015,18 @@ def tsem(a, limits=None, inclusive=(True, True), axis=0, ddof=1):
     """
-    Calculates the nth moment about the mean for a sample.
+    Calculates the nth central moment about the mean for a sample.
+
```
Central moment is by definition the moment about the mean. I'd only leave on of qualifiers.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1015,16 +1015,18 @@ def tsem(a, limits=None, inclusive=(True, True), axis=0, ddof=1):
     a : array_like
        data
-    moment : int, optional
+    moment : int or array_like of ints, optional
```
Worth stating the default value.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -1037,9 +1039,57 @@ def moment(a, moment=1, axis=0):
+
+    Notes
+    -----
+    [1]_ The k-th central moment of a data sample is:
```
Reference is a leftover from some rewriting?

aeklant commented:
Fixed the latest doc observations. Please review.

rgommers commented:
All comments addressed, this looks good now. So in it goes. Thanks @aeklant, @ev-br
